Libraries/Opc.Ua.Server/Subscription/MonitoredItem/MonitoredItemQueue.cs

```diff
@@ -0,0 +1,426 @@
+            {
+                if (m_values == null || m_start == -1)
+                {
+                    return 0;
+                }
+
+                if (m_start < m_end)
+                {
+                    return m_end - m_start;
+                }
+
+                return m_values.Length - m_start + m_end;
+            }
+        }
+        /// <inheritdoc/>
+        public bool IsDurable => m_isDurable;
+
+
+        /// <summary>
+        /// Adds the value to the queue.
+        /// </summary>
+        /// <param name="value">The value to queue.</param>
+        /// <param name="error">The error to queue.</param>
+        public void Enqueue(DataValue value, ServiceResult error)
+        {
+            if (m_values == null)
+            {
+                throw new InvalidOperationException("Cannot enqueue Value. Queue size not set.");
+            }
+
+            //check for full queue
+            if (ItemsInQueue == m_values.Length)
+            {
+                Dequeue(out _, out _);
+            }
+
+            // check for empty queue.
+            if (m_start < 0)
+            {
+                m_start = 0;
+                m_end = 1;
+
+                m_values[m_start] = value;
+
+                if (m_errors != null)
+                {
+                    m_errors[m_start] = error;
+                }
+
+                return;
+            }
+
+            int next = m_end;
+
+            // check for wrap around.
+            if (next >= m_values.Length)
+            {
+                next = 0;
+            }
+
+            // add value.
+            m_values[next] = value;
+
+            if (m_errors != null)
+            {
+                m_errors[next] = error;
+            }
+
+            m_end = next + 1;
+        }
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            //only needed for unmanaged resources
+        }
+
+        /// <inheritdoc/>
+        public void OverwriteLastValue(DataValue value, ServiceResult error)
+        {
+            if (m_values == null)
+            {
+                throw new InvalidOperationException("Cannot write Value. Queue size not set.");
+            }
+
+            int last = m_end - 1;
+
+            if (last < 0)
+            {
+                last = m_values.Length - 1;
+            }
+
+            // replace last value and error.
+            m_values[last] = value;
+
+            if (m_errors != null)
+            {
+                m_errors[last] = error;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void SetQueueSize(uint queueSize, bool queueErrors)
+        {
+            int length = (int)queueSize;
+
+            // create new queue.
+            DataValue[] values = new DataValue[length];
+            ServiceResult[] errors = null;
+
+            if (queueErrors)
+            {
+                errors = new ServiceResult[length];
+            }
+
+            // copy existing values.
```

Comment:
Use same logic as for events to do the copy (discard oldest /newest is currently not implemened)